Replace Track Alias with Publisher-Chosen Subscription ID (#1)
afrind/moq-transport: draft-ietf-moq-transport.md

UTF-8. Any specification that does needs to specify the canonicalization into
the bytes in the Track Namespace or Track Name such that exact comparison works.

### Subscription Identifiers {#subscription-id}

To optimize wire efficiency, some MoQT messages refer to a track by a numeric
identifier, rather than the Full Track Name. When a subscriber makes a SUBSCRIBE
or FETCH request for a Track, it associates a Request ID with the request.  The
Request ID (RSID) can be used as a Subscription ID by shifting it left by one,

> eg: `Request Subscription ID = Request ID << 1`.

Reviewer: I am still not clear, what RequestID is used to generate RSID . Also when should one use which (ReqID vs RSID).

Author: SUBSCRIBE contains a Request ID, as do (almost) all Control Messages, since we landed 905.  That said, you will never send Request ID *directly* on the wire after SUBSCRIBE/SUBSCRIBE_OK/SUBSCRIBE_ERROR as this PR is currently written - all other operations use Subscription ID.  You can scan the Message formats below to see if they use Request ID or Subscription ID.

MoQT also allows the publisher to choose an identifier for each subscription,
called the Publisher-Chosen Subscription ID (PSID). PSIDs are odd and have a
maximum value of 2^62-1. The same identifier MUST NOT be assigned to different

> Tracks in the same session.  The publisher is not required to specify a PSID.

Reviewer: As i mentioned in the slack chat, I think we should not make it optional

Author: Noted, I will add a DISCUSS here since this will need to go to the full WG.

> The term Subscription ID refers to either a Request Subscription ID or

Reviewer: we should rename Request Subscription ID to Subscription Request ID, since the commonality of the term Subscription ID is confusing.

Author: Hah, I just switched it from Subscription Request ID to Request Subscription ID, but I liked it the original way better.  I will switch it back.

Publisher-Chosen Subscription ID, determined by the least significant bit of
the identifier.  Subgroups, Datagrams as well as control messages for
established subscriptions such as SUBSCRIBE_UPDATE, UNSUBSCRIBE and
SUBSCRIBE_DONE use Subscription IDs.  The sender chooses the identifier type

> based on its preference.

Reviewer: I will remove the last sentence as it gets more clarified in the relay section OR i would also add reference to the relay section here

UTF-8. Any specification that does needs to specify the canonicalization into
the bytes in the Track Namespace or Track Name such that exact comparison works.

### Subscription Identifiers {#subscription-id}

To optimize wire efficiency, some MoQT messages refer to a track by a numeric
identifier, rather than the Full Track Name. When a subscriber makes a SUBSCRIBE
or FETCH request for a track, it associates a Request ID with the request.  The

> corresponding Subscription Request ID is equal to `Request ID << 1`.

Reviewer: there is no Susbcription Request ID. So may be we should remove the word "Subscription"

Or what is the difference between RequestID and Subscription Request ID. Where is the latter used ?

Author: I'm defining Subscription Request ID here.  It can be used anywhere to refer to this subscription (control messages and data streams).

Reviewer 2: I would expect it to be called something like "Subscriber-Chosen Subscription ID", except of course, the subscriber doesn't have the liberty of actually choosing it as far as I can tell.

Also how does this work in PUBLISH world?

Author: > I would expect it to be called something like "Subscriber-Chosen Subscription ID", except of course, the subscriber doesn't have the liberty of actually choosing it as far as I can tell.

Right that didn't seem like a good name.  I'm open to something else, it's an identifier for the subscription based on the request ID that initiated it.  Alternately, we can say the 0 bit is not part of the ID, and you use >> to recover the original Request ID.

It works symmetrically for publish, because Publish also includes a Request ID.

Reviewer 2: > It works symmetrically for publish, because Publish also includes a Request ID.

Interesting.  So the publisher effectively just has two IDs it knows at the moment it sends a PUBLISH message, and it can pick either?  I am not sure I see the justification for having two separate IDs.

Author: >  So the publisher effectively just has two IDs it knows at the moment it sends a PUBLISH message, and it can pick either? 

There _may_ be only one ID -- PSID is not required for publish.  You can publish using Subscription Request ID too, it just has the same control and data reordering problems that PSID does.

The justification for using Request ID in Publish is so a receiver can limit the number of publishes using MAX_REQUEST_ID.

Author: I renamed it `Request Subscription ID`, and tried to make it clear that this is a new term, and that Subscription ID = RSID or PSID

MoQT also allows the publisher to choose an identifier for each track, called

> the Publisher-Chosen Track ID (PTID). PTIDs are odd and have a maximum

Author: ```suggestion
the Publisher-Chosen Subscription ID (PSID). PSIDs are odd and have a maximum
```

> the Publisher-Chosen Subscription ID (PSID). PSIDs are odd and have a maximum

Reviewer: ```suggestion
the Publisher-Chosen Track ID (PTID). PTIDs are odd and have a maximum
```

Reviewer: I think calling it a publisher chosen subscription id is confusing when a relay wants to use same ID for all the subscribers for a given track.

Reviewer 2: There is only one instance of a track within a single MoQ connection.  In fact, if you publish the same track twice within a single MoQ connection, I imagine you would have to use a different ID due to uniqueness requirement.

Author: Ian has a PR about allowing multiple subscribes to the same track as long as they have non-overlapping filters (774), which allows the receiver to deterministically map the objects received back to the the subscription that generated it.

That said, there's no _requirement_ that the original publisher assign the same PSID to different subscriptions to the same track, so I think PSID is fine, though I should change 

> MoQT also allows the publisher to choose an identifier for each track

to

> MoQT also allows the publisher to choose an identifier for each subscription

> value of 2^61-1. The same identifier MUST NOT be assigned multiple tracks in the

Reviewer: ```suggestion
value of 2^61-1. The same identifier MUST NOT be assigned to multiple tracks in the
```

> value of 2^61-1. The same identifier MUST NOT be assigned to multiple tracks in the

Reviewer 2: Wouldn't they have the maximum value of 2^62-1?

Author: You are right - since I say they are odd.

> value of 2^61-1. The same identifier MUST NOT be assigned to multiple tracks in the

Author: ```suggestion
value of 2^62-1. The same identifier MUST NOT be assigned to multiple tracks in the
```

> same session.  PSIDs are required when the publisher initiates the subscription

Reviewer: This is not totally true. These IDs can be chosen even when the publisher wants to publish objects in response to a subscription

Author: This is part of the optionality discussion below.

(TODO: delete this; PUBLISH message is forthcoming) and are optional when the
subscriber initiates it.

When a publisher sends Subgroups or Datagrams, these can contain either a
Subscription Request ID or Publisher-Chosen Subscription ID, determined by the

> least significant bit of the identifier. Control messages for an established

Reviewer: Control message needs to have both exchanged. If not, the subscriber will not be able to know which PSID corresponds to which track.
Say if Control message used RSID (8) for TrackA and object messages uses PSID (33) .. How can the subscriber know 33 is for TrackA ?

Author: The word `established` indicates that the subscriber has seen SUBSCRIBE_OK.

subscription can also use either identifier.  Messages that control the
subscription such as SUBSCRIBE_UPDATE, UNSUBSCRIBE and SUBSCRIBE_DONE also use

> Subscription IDs.

Reviewer 2: This seems to be a downgrade from the current draft, which only allows Subscribe ID in those.

Author: I can require this to be Request ID, but it seemed reasonable to be symmetric.

UTF-8. Any specification that does needs to specify the canonicalization into
the bytes in the Track Namespace or Track Name such that exact comparison works.

### Subscription Identifiers

To optimize wire efficiency, some MoQT messages refer to a track by a numeric
identifier, rather than the Full Track Name. When a subscriber makes a SUBSCRIBE
or FETCH request for a track, it associates a Subscriber-Chosen Subscription ID
(SSID) with the request. SSIDs are scoped to a single MoQT session.  They are
even; they start at 0 and increase by 2 and have a maxmium value of 2^62-2.  The
publisher limits the number of requests a subscriber can make using the
MAX_SUBSCRIPTION_ID SETUP parameter or MAX_SUBSCRIPTION_ID control message.

MoQT also allows the publisher to choose an identifier for each track, called
the Publisher-Chosen Subscription ID (PSID). PSIDs are odd and have a maximum
value of 2^61-1. The same identifier MUST NOT be assigned multiple tracks in the
same session.  PSIDs are required when the publisher initiates the subscription

> (TODO: delete this; PUBLISH message is forthcoming) and are optional when the

Reviewer 3: Making it optional is ok, but it seems cleaner if every subscription has both, regardless of who initiated it.

Author: I'm thinking the optional nature makes it easier for publishers who don't care about this.

Reviewer: I am inclined of keeping it as non optional as well.

Author: There is a contingent of folks in the working group who do not want this feature as it is not necessary in their application.  Only publishers that have such an ID should be need to set it.  We can mark this as an open question in the PR.

Reviewer: we are using this in control message and as long as we do good job at clarifying its purpose, we should be fine. We can always add it and make it optional, if people think otherwise too.

Author: I rewrote this bit to say that PSID is explicitly optional.  Will add to a list of items to discuss further.

subscriber initiates it.

When a publisher sends Subgroups or Datagrams, these can contain either a

> Subscriber-Chosen or Publisher-Chosen Subscription ID, determined by the LSB of

Reviewer 3: Can you spell out least significant bit here.  I suspect most people would figure it out, but I'd rather use the long form once before using the abbreviation.

* Protocol Violation: The remote endpoint performed an action that was
  disallowed by the specification.

* Duplicate Subscription ID: The endpoint attempted to use a Publisher-Chosen

>   Subscription ID that was already in use.

Reviewer: This is very confusing. Can there be duplicates for RSID as well as PSID. May be we need 2 errors
Duplicate PSID
Duplicate RSID 
??

Author: That has a different error code "Invalid Request ID" (which is missing its description here)

> Invalid Request ID: The session was closed because the endpoint used a Request ID that was smaller than or equal to a previously received request ID, or the least-significant bit of the request ID was incorrect for the endpoint.

Reviewer: This is incorrect . From what i see, Subscription ID is either a PSID or RSID. This error says the error is only scoped to PSID.

Author: A duplicate Request ID is only an _error_ when it comes in a request message (eg: SUBSCRIBE, ANNOUNCE, etc).  Invalid Request ID is meant to capture that.  There's not really such a thing as a duplicate RSID - these only appear on the data plane, in SUBSCRIBE_DONE, and potentially UNSUBSCRIBE/SUBSCRIBE_UPDATE.  Duplicates are never an error in those places.

So this error code is only meant to capture duplicate PSIDs, we can update the name to reflect that.

each active upstream subscription that matches that namespace, it SHOULD send a
SUBSCRIBE to the publisher that sent the ANNOUNCE.

If a relay receives a Publisher-Chosen Subscription ID in SUBSCRIBE_OK, it
SHOULD assign this ID to downstream subscriptions for the same track. Since
subscribers can request tracks from uncoordinated publishers through a single
relay session, it is not always possible to reuse the upstream PSID.  If there
is no upstream Publisher-Chosen Subscription ID, or if the upstream PSID is
already in use downstream for a different track, the relay SHOULD NOT set a
Publisher-Chosen Subscription ID, and use the Subscriber-Chosen Subscription ID

> instead.

Reviewer 3: Do we need this paragraph?

Author: I think so - we need to explain that relays should be trying to reuse upstream IDs on the downstream side, and also that this doesn't always work.  This is another reason to make PSID optional - I'd rather just use SSID on the downstream than have the relay invent a new PSID, that itself might conflict with future ones.

Reviewer 3: That means a subscriber can't limit how many PUBLISHes a peer sends it if they're not monotonically increasing.  I'm not sure I want to give up on that so quickly unless we really think it's not useful?

Author: Fixed this by rebasing on Request ID PR -- PUBLISH now has a Request ID and the subscriber can limit the total number of requests/publishes.

each active upstream subscription that matches that namespace, it SHOULD send a
SUBSCRIBE to the publisher that sent the ANNOUNCE.

If a relay receives a Publisher-Chosen Subscription ID in SUBSCRIBE_OK, it

> SHOULD assign this ID to downstream subscriptions for the same track. Since

Reviewer 2: But why, though?

Author: Are you saying you want more text explaining why relays SHOULD preserve these IDs, or you saying you don't want any normative requirement for relays with respect to the IDs?

Reviewer 2: > more text explaining why relays SHOULD preserve these IDs

Well, that would be a good start.

> you saying you don't want any normative requirement for relays with respect to the IDs?

I have no idea as to what we're trying to accomplish here, so I can't say either way.

Author: I changed this from SHOULD to MAY.  The purpose is for systems that have such identifiers to convey them end-to-end.

subscribers can request tracks from uncoordinated publishers through a single
relay session, it is not always possible to reuse the upstream PSID.  If there
is no upstream Publisher-Chosen Subscription ID, or if the upstream PSID is

> already in use downstream for a different track, the relay SHOULD NOT set a

Reviewer: may be it will help clarify further on where the relay SHOULD NOT set ? From what I understand, it is set on the downstream subscription?

Author: I think the conditions are pretty clear:

1. There is no upstream Publisher-Chosen Subscription ID
2. The upstream PSID is already in use downstream for a different track

Publisher-Chosen Subscription ID, and use the Subscription Request ID instead.

> Relays use the Subscription Identifier (see {{subscription-id}}) in an incoming

Reviewer: ```suggestion
Relays use the Subscription Request Identifier (see {{subscription-id}}) in an incoming
```

Reviewer: This assumes that OBJECT message always must use RSID. I thought they can  contain PSID alternatively.

As part of control message exchange, relay will know if a PSID/RSID is expected for a track and uses that to map to right subscribers. Isn't that right ?

Author: > This assumes that OBJECT message always must use RSID. I thought they can contain PSID alternatively.

"Subscription Identifier" refers to _either_ PSID or RSID.

> As part of control message exchange, relay will know if a PSID/RSID is expected for a track and uses that to map to right subscribers. Isn't that right ?

Every receiver, including a relay, needs to be able to receive a Subgroup/Datagram with either ID and use that to look up subscription state.

Note that since Setup parameters use a separate namespace, it is impossible for
these parameters to appear in Setup messages.

### PUBLISHER CHOSEN SUBSCRIPTION ID {#publisher-chosen-subscription-id}

The PUBLISHER CHOSEN SUBSCRIPTION ID parameter (Parameter Type 0x6) contains the
PSID for this subscription (see {{subscription-id}}).  It can appear in
SUBSCRIBE_OK.  If omitted, this subscription does not have a PSID and the

> publisher will only publish objects using the Subscription Request ID.

Author: ```suggestion
publisher will only publish objects using the Request Subscription ID.
```

Note that since Setup parameters use a separate namespace, it is impossible for
these parameters to appear in Setup messages.

### PUBLISHER CHOSEN SUBSCRIPTION ID {#publisher-chosen-subscription-id}

The PUBLISHER CHOSEN SUBSCRIPTION ID parameter (Parameter Type 0x6) contains the
PSID for this subscription (see {{subscription-id}}).  It can appear in

> SUBSCRIBE_OK.  If omitted, this subscription does not have a PSID and the

Reviewer: ```suggestion
SUBSCRIBE_OK, SUBGROUP, Datagram.  If omitted, this subscription does not have a PSID and the
```

Author: This is a Parameter to it's identifying the control messages that are valid.

Reviewer: I don't think this should be in parameter.

  If a server receives a GOAWAY with a non-zero New Session URI Length it MUST
  terminate the session with a Protocol Violation.

## MAX_SUBSCRIPTION_ID {#message-max-subscribe-id}

A publisher sends a MAX_SUBSCRIPTION_ID message to increase the number of
subscriptions a subscriber can create within a session.

The Maximum Subscriber-Chosen Subscription ID MUST only increase within a
session, and receipt of a MAX_SUBSCRIPTION_ID message with an equal or smaller
Subscription ID value is a 'Protocol Violation'.

~~~
MAX_SUBSCRIPTION_ID
{
  Type (i) = 0x15,
  Length (i),
  Subscription ID (i),
}
~~~
{: #moq-transport-max-subscribe-id format title="MOQT MAX_SUBSCRIPTION_ID Message"}

> * Subscription ID: The new Maximum Subscriber-Chosen Subscription ID for the

Reviewer 3: So this limits subscribers, but not publishers?  Do we want to use even/odd here to differentiate the limits like we do in QUIC?

Author: You can't do that because PSID is not restricted to increase - they can be sprayed all over the 61 bit space.  We could limit control messages using something like control message IDs (moq-transport PR 898), but I'm not sure how we can limit publisher initiated subscriptions.

{
  Type (i) = 0x4,
  Length (i),

>   Subscription ID (i),

Reviewer 3: I don't have a strong preference, but renaming Subscribe ID to Subscription ID makes this PR quite a bit larger.

Author: Fair, I can move that to a subsequent PR.  Since subscriptions can now be established by SUBSCRIBE or PUBLISH, Subscription ID seems more appropriate.

* Subscribe Parameters: The parameters are defined in {{version-specific-params}}.

> PUBLISHER_CHOSEN_SUBSCRIPTION_ID: The Publisher-Chosen Subscription ID for

Reviewer 3: What field is this referencing and how would it be omitted?

Author: It's intended to be a parameter, so it can be easily optional.  I don't love how we have parameter defs miles away from where they are used.

Reviewer 3: Ok, I now understand what you're trying to do, but given right above it says all parameters are defined elsewhere, this was confusing for me.  Most parameters are shared among messages, and this one isn't, which might be an argument for it being a field?  Other control messages can reference this ID, but can't change it.

~~~
{: #moq-transport-subscribe-update-format title="MOQT SUBSCRIBE_UPDATE Message"}

> * Subscription ID: Either the Subscription Request ID or Publisher-Chosen

Author: ```suggestion
* Subscription ID: Either the Request Subscription ID or Publisher-Chosen
```

UNSUBSCRIBE Message {
  Type (i) = 0xA,
  Length (i),
  Subscription ID (i)
}
~~~
{: #moq-transport-unsubscribe-format title="MOQT UNSUBSCRIBE Message"}

> * Subscription ID: Either the Subscription Request ID or Publisher-Chosen

Author: ```suggestion
* Subscription ID: Either the Request Subscription ID or Publisher-Chosen
```

~~~
{: #moq-transport-subscribe-fin-format title="MOQT SUBSCRIBE_DONE Message"}

> * Subscription ID: Either the Subscription Request ID or Publisher-Chosen

Author: ```suggestion
* Subscription ID: Either the Request Subscription ID or Publisher-Chosen
```

~~~
{: #object-datagram-status-format title="MOQT OBJECT_DATAGRAM_STATUS"}

* Subscription ID: the Subscriber or Publisher chosen Subscription Identifier
  (see {{subscription-id}}) indicating the subscription this Datagram belongs

>   to.  If an endpoint receives a datagram with a Subscription Request ID that

Author: ```suggestion
  to.  If an endpoint receives a datagram with a Request Subscription ID that
```

### Subgroup Header

When a stream begins with `SUBGROUP_HEADER`, all Objects on the stream
belong to the track requested in the subscription identified by `Subscription
Identifier` and the subgroup indicated by 'Group ID' and `Subgroup ID`.

~~~
SUBGROUP_HEADER {
  Subscription ID (i),
  Group ID (i),
  Subgroup ID (i),
  Publisher Priority (8),
}
~~~
{: #object-header-format title="MOQT SUBGROUP_HEADER"}

* Subscription ID: the Subscriber or Publisher chosen Subscription
  Identifier indicating the subscription this Subgroup belongs to. If an

>   endpoint receives a subgroup with aSubscriber-Chosen Subscription

Reviewer 3: ```suggestion
  endpoint receives a subgroup with a Subscriber-Chosen Subscription
```

### Subgroup Header

When a stream begins with `SUBGROUP_HEADER`, all Objects on the stream
belong to the track requested in the subscription identified by `Subscription
Identifier` and the subgroup indicated by 'Group ID' and `Subgroup ID`.

~~~
SUBGROUP_HEADER {
  Subscription ID (i),
  Group ID (i),
  Subgroup ID (i),
  Publisher Priority (8),
}
~~~
{: #object-header-format title="MOQT SUBGROUP_HEADER"}

* Subscription ID: the Subscriber or Publisher chosen Subscription Identifier
  (see {{subscription-id}}) indicating the subscription this Subgroup belongs

>   to. If an endpoint receives a subgroup with a Subscription Request ID that

Author: ```suggestion
  to. If an endpoint receives a subgroup with a Request Subscription ID that
```
